src/main/ruby/truffleruby/core/posix.rb

```diff
@@ -425,11 +425,62 @@ def self.read_string_native(io, length)
     end
   end
 
-  def self.read_to_buffer_native(io, length)
+  def self.execute_posix_read(io, buffer, length)
+    fd = io.fileno
+    return Truffle::POSIX.read(fd, buffer, length) unless io.timeout
+
+    deadline = Time.now.to_f + io.timeout
```

Comment:
`Process.clock_gettime(Process::CLOCK_MONOTONIC` should be used for anything timeout-related, Time.now is affected by system time changes/adjustments

Reply:
Thanks, will update.